cygnss_wetlands/cygnss/aggregate.py

```diff
@@ -0,0 +1,148 @@
+        # handles rows
+        if r == 0:
+            rows.append(r + n)
+        elif r == grid.n_rows - 1:
+            rows.append(r - n)
+        else:
+            rows.append(r + n)
+            rows.append(r - n)
+
+        # handle cols
+        if c == 0:
+            cols.append(grid.n_cols - n)
+            cols.append(c + n)
+        elif c == grid.n_cols - 1:
+            cols = cols.append(n - 1)
+            cols.append(c - n)
+        else:
+            cols.append(c + n)
+            cols.append(c - n)
+
+    rows_comb, cols_comb = zip(*itertools.product(rows, cols))
+    return np.array(rows_comb).astype(np.int16), np.array(cols_comb).astype(np.int16)
+
+
+def inverse_distance(data: pd.DataFrame, grid: GenericGrid, variable_name: str) -> np.ndarray:
+    """
+    This algorithm finds the nearest neighbors in all 8-directions. The values are weighted
+        by distance, and can use a linear, squared, or other weighting to apply greater
+        weights to points close to the grid intersection.
+
+    Caveats:
+        - For now, we ONLY support squared weights
+        - Only supports grid CRS that has meters as base unit
+    """
+    # Ensure selected grid CRS units is meters
+    try:
+        assert grid.proj.crs.axis_info[0].unit_name == "metre"
+    except AssertionError:
+        raise Exception(
+            "Need to either add support for other units - or add ability accept different spellings, "
+            "meter / metre / m  for given CRS?)"
+        )
+
+    # to populate a gridcell, only consider values within this distance threshold
+    dist_threshold_m = CYGNSS_IDW_DISTANCE_THRESHOLD_KM * 1000  # in meters
+
+    # Get row, cols for grid
+    col_row = data.apply(lambda x: grid.lonlat2cr(x.sp_lon, x.sp_lat), axis=1).values
+    data[["col", "row"]] = pd.DataFrame(col_row.tolist(), index=data.index)
+
+    # initialize empty grids to start with
+    sum_grid = np.full((grid.n_rows, grid.n_cols), 0, dtype=np.float32)
+    count_grid = np.full((grid.n_rows, grid.n_cols), 0, dtype=np.float32)
+
+    # iterate over each data value - find neighbors, accumulate weighted values and counts
+    # todo: is there a more efficient way to do this?
+    for idx, row in data.iterrows():
+        neighbors_row, neighbors_col = get_nearby_gridcells(grid, row.row, row.col)
+        neighbors_lon, neighbors_lat = zip(
+            *[grid.cr2lonlat(neighbors_col[i], neighbors_row[i]) for i in range(len(neighbors_row))]
+        )
+        dist_to_swathpt = haversine(
+            lat1=neighbors_lat,
+            lon1=neighbors_lon,
+            lat2=np.repeat(row.sp_lat, len(neighbors_lat)),
+            lon2=np.repeat(row.sp_lon, len(neighbors_lon)),
+        )
+        mask = dist_to_swathpt < dist_threshold_m
+        sum_grid[neighbors_row[mask], neighbors_col[mask]] += row[variable_name] / (dist_to_swathpt[mask] ** 2)
+        count_grid[neighbors_row[mask], neighbors_col[mask]] += 1.0 / (dist_to_swathpt[mask] ** 2)
```

Comment:
Updated the `haversinve()` function to avoid 0 distances (if specified) by adding a small nominal amount in cases where points are identical. But you're right.. this is suuuuper unlikely where a specular point would fall exactly on a gridcell centroid